src/action.py

```diff
@@ -1,30 +1,79 @@
 
 import glob
 import click
-import os
+import re
+
+from sys import exit
+from os import listdir
+from os.path import join, dirname, relpath, isfile, isdir
 from models.config import Config
 from collections import defaultdict
 from itertools import chain
 
+#TODO DRY up the file leader - only define '/var/lib/adminware/tools' once
 class ClickGlob:
-    def __glob_configs(namespace):
-        parts = ['/var/lib/adminware/tools',
-                 namespace, '*/config.yaml']
-        paths = glob.glob(os.path.join(*parts))
-        return list(map(lambda x: Config(x), paths))
+    def __glob_dirs(namespace):
+        parts = ['/var/lib/adminware/tools', namespace, '*']
+        paths = glob.glob(join(*parts))
+        return paths
+
+    # decorator for __glob_dir to get all valid paths (ending in config.yaml) at once
+    def __glob_all(globbing_func):
+        def __wrapper(*args, **kwargs):
+            def __sub_wrapper(*args, **kwargs):
+                paths = globbing_func(*args)
+                for path in paths:
+                    if isfile('{}/config.yaml'.format(path)):
+                        kwargs['all_paths'] += [join(path, 'config.yaml')]
+                    else:
+                        namespace = path[len('/var/lib/adminware/tools/'):]
+                        __sub_wrapper(namespace, **kwargs)
+                return kwargs['all_paths']
+
+            kwargs['all_paths']=[]
+            all_paths = __sub_wrapper(*args, **kwargs)
+            return all_paths
```

Comment:
Does python allow for?
```
return __sub_wrapper(*args, **kwargs)
```

Reply:
also done here https://github.com/alces-software/adminware/pull/87/commits/be89624fbf4de21d3f15fe67aaae6a0530be8a56